src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs

```diff
@@ -369,39 +357,46 @@ private void ProcessMessageCore(IncomingMessage message)
-                    try
+                    using (CancellationTokenSource cancel = new CancellationTokenSource())
                     {
-                        this.ConnectedNetworkPeers.Add(networkPeer);
-                        networkPeer.StateChanged += Peer_StateChanged;
-                        networkPeer.RespondToHandShake(cancel.Token);
-                    }
-                    catch (OperationCanceledException)
-                    {
-                        this.logger.LogTrace("Remote peer haven't responded within 10 seconds of the handshake completion, dropping connection.");
-
-                        networkPeer.DisconnectAsync();
-
-                        this.logger.LogTrace("(-)[HANDSHAKE_TIMEDOUT]");
-                        throw;
-                    }
-                    catch (Exception ex)
-                    {
-                        this.logger.LogTrace("Exception occurred: {0}", ex.ToString());
-
-                        networkPeer.DisconnectAsync();
-
-                        this.logger.LogTrace("(-)[HANDSHAKE_EXCEPTION]");
-                        throw;
+                        cancel.CancelAfter(TimeSpan.FromSeconds(10.0));
+                        try
+                        {
+                            this.ConnectedNetworkPeers.Add(networkPeer);
+                            networkPeer.StateChanged += Peer_StateChanged;
+                            networkPeer.RespondToHandShake(cancel.Token);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            this.logger.LogTrace("Remote peer haven't responded within 10 seconds of the handshake completion, dropping connection.");
+
+                            networkPeer.DisconnectAsync();
+
+                            this.logger.LogTrace("(-)[HANDSHAKE_TIMEDOUT]");
+                            throw;
+                        }
+                        catch (Exception ex)
+                        {
+                            this.logger.LogTrace("Exception occurred: {0}", ex.ToString());
+
+                            networkPeer.DisconnectAsync();
+
+                            this.logger.LogTrace("(-)[HANDSHAKE_EXCEPTION]");
+                            throw;
+                        }
                     }
                 }
             }
 
-            this.MessageReceived?.Invoke(this, message);
             this.logger.LogTrace("(-)");
         }
 
+        /// <summary>
+        /// Callback that is called when a network status of a connected client peer changes.
+        /// <para>The peer is removed from the list of connected peers if the connection has been terminated for any reason.</para>
+        /// </summary>
+        /// <param name="peer">The connected peer.</param>
+        /// <param name="oldState">Previous state of the peer. New state of the peer is stored in its <see cref="NetworkPeer.State"/> property.</param>
```

Comment:
Perhaps the server should have its own behaviour for add/remove.
Just something to consider

Reply:
my strategy now is to make these guys as slim as possible and remove all the stuff we don't use, then understand it what is actually there that remained 

and ONLY THEN to somehow change the design, add new features etc.

so in general, yes, anything is possible then, but it will take time to get there

Reply:
in other words - if you have an idea, add it to the issue, so that we don't forget about it